doc/developer/design/20230602_default_privileges.md

```diff
@@ -0,0 +1,144 @@
+  reserved for all roles.
+- `schema_id text`: The id of the schema this default privilege applies to or NULL.
+- `database_id text`: The id of the database this default privilege applies to or NULL.
+- `object_type char`: Type of object this default privilege applies to.
+- `privileges mz_aclitem[]`: The default privileges to apply.
+
+If both `schema_id` and `database_id` are NULL, then the default privilege applies to all objects of
+a certain type in all databases and schemas.
+
+`object_type` can be one of the following characters:
+
+- `n`: schema
+- `d`: database
+- `C`: cluster
+- `r`: relation
+- `T`: type
+- `s`: secret
+- `c`: connection
+
+`mz_default_privileges` will be pre-populated with certain system default privileges. Currently, the
+only one is:
+
+- `(<id>, <reserved-role-id-for-all-roles>, NULL, NULL, T, {=U/mz_system})`. i.e. `PUBLIC` is
+  granted `USAGE` on every type.
+
+`ALTER DEFAULT PRIVILEGES` is a command that will have the following syntax (formatted the same way
+that PostgreSQL formats SQL syntax in their documentation):
+
+`ALTER DEFAULT PRIVILEGES [ FOR <role_specification> ] [ { IN SCHEMA <schema_name> [, ...] | IN DATABASE <database_name> [, ...] } ] <abbreviated_grant_or_revoke>`
+
+`<role_specification>: FOR { { ROLE | USER } <target_role> [, ...] | ALL ROLES }`
+
+`<abbreviated_grant_or_revoke>`: `{ <grant> | <revoke> }`
+
+`<grant>`: `GRANT { <privilege> [, ...] | ALL [ PRIVILEGES ] } ON <object_type> TO [ GROUP ] <role_name> [, ...]`
+
+`<revoke>`: `REVOKE { <privilege> [, ...] | ALL [ PRIVILEGES ] } ON <object_type> FROM [ GROUP ] <role_name> [, ...]`
+
+The grant variant will add rows to `mz_default_privileges`, such that:
+
+- `role_id` is filled in from `<role_specification>`.
+- `schema_id` is filled in from `<schema_name>`.
+- `oject_type` is filled in from `<object_type>`.
+- `privilege` is filled in from `<privilege>` and `<role_name>`.
+
+If a matching row already exists, then the `mz_aclitem` will be updated instead of having a new row
+added.
+
+The revoke variant will update an existing row, if one exists. Otherwise, it will have no affect. If
+a row ends up with an empty `mz_aclitem` then it will be removed.
+
+Users must be a superuser to specify the `ALL ROLES` role specification. Users must be a member of
+both `<target_role>` and `<role_name>`.
+
+When an object is first created materialize will do the following in order:
+
+1. Grant the owner all privileges on that object.
+2. Grant any privileges that match a privilege in the `mz_default_privileges` table on that object.
+
+### Differences from PostgreSQL
+
+This section will highlight the differences between PostgreSQL default privileges and the proposed
+default privileges for Materialize.
+
+- This proposal has no special system privileges that are hard-coded into the system. Instead, we
+  pre-populate the user default privileges table with some initial default privileges.
```

Comment:
IMO we should 100% do this, even though it's a deviation from Postgres. To me it seems like Postgres has two sets of "defaults", a hidden set that is hard coded and the ones visible in `pg_default_acl`, which is very confusing to someone who doesn't know about privileges in-depth.

Out of curiosity is there precedence for this design? What do other Postgres compatible databases do, and is there any chatter in the mailing list about why Postgres has this design?

Reply:
The `ALL ROLES` syntax was inspired by CockroachDB (https://www.cockroachlabs.com/docs/stable/alter-default-privileges.html)

Looking through their docs (https://www.cockroachlabs.com/docs/v23.1/security-reference/authorization#default-privileges), they never mention these hard coded system defaults. Also I've never run `ALTER DEFAULT PRIVILEGE` on my cockroach cluster, yet I get the following results:

```
root=> SELECT * FROM crdb_internal.default_privileges LIMIT 10;
 database_name | schema_name | role  | for_all_roles | object_type | grantee | privilege_type | is_grantable 
---------------+-------------+-------+---------------+-------------+---------+----------------+--------------
 defaultdb     |             | admin | f             | tables      | admin   | ALL            | t
 defaultdb     |             | admin | f             | sequences   | admin   | ALL            | t
 defaultdb     |             | admin | f             | types       | admin   | ALL            | t
 defaultdb     |             | admin | f             | schemas     | admin   | ALL            | t
 defaultdb     |             | admin | f             | functions   | admin   | ALL            | t
 defaultdb     |             | admin | f             | types       | public  | USAGE          | f
 defaultdb     |             | root  | f             | tables      | root    | ALL            | t
 defaultdb     |             | root  | f             | sequences   | root    | ALL            | t
 defaultdb     |             | root  | f             | types       | root    | ALL            | t
 defaultdb     |             | root  | f             | schemas     | root    | ALL            | t
(10 rows)

```

Note that the `USAGE` privilege on `types` to `PUBLIC` is one of the PostgreSQL system hard-coded default privilege. So it looks like CockroachDB takes the same or very similar approach to what's proposed here.

Reply:
Agree this is good. Do we need to make sure users can't accidentally get themselves into a perma-broken state? Or are we always safe because frontegg defines superusers who can always do anything?

Reply:
We can't get into a perma-broken state from default privileges, because they will only allow adding more privileges, they can't take privileges away. The pre-populated privileges, like `USAGE` on all types to `PUBLIC`, is a nice to have but not strictly necessary.

In general though, yes, superusers will help avoid perma-broken states because they can always do anything. Additionally, every object must always have a non-deleted object owner who can always grant/revoke any privilege on that object.